File: ToOffer/09-QueueStack/StackWithTwoQueues/Queue.m
//
//  Queue.m
//  StackWithTwoQueues
//
//  Created by Csy on 2018/11/5.
//  Copyright © 2018 Csy. All rights reserved.
//

#import "Queue.h"

@interface Queue ()

@property (nonatomic, strong) NSMutableArray *array;

@end

@implementation Queue

- (instancetype)init {
    if (self = [super init]) {
        _array = [NSMutableArray array];
    }
    return self;
}

- (void)enqueue:(id)object {
    if (object == nil) {
        return;
    }
    [self.array addObject:object];
}

- (id)dequeue {
    id object = [self.array firstObject];
    if (object) {
        [self.array removeObject:object];
    }
    return object;
}

- (BOOL)isEmpty {
    return (self.array.count == 0);
}

- (NSInteger)count {
    return self.array.count;
}

@end

File: LeetCode/StackQueue/StackQueue/Queue.swift
//
//  Queue.swift
//  StackQueue
//
//  Created by Csy on 2019/2/14.
//  Copyright © 2019 CXL. All rights reserved.
//

import Foundation

public struct Queue<T> {
    fileprivate var array = [T]()
    
    public var count: Int {
        return array.count
    }
    
    public var isEmpty: Bool {
        return array.isEmpty
    }
    
    public mutating func enqueue(_ element: T) {
        array.append(element)
    }
    
    public mutating func dequeue() -> T? {
        if isEmpty {
            return nil
        } else {
            return array.removeFirst()
        }
    }
    
    public var front: T? {
        return array.first
    }
}

/*
 Deque (pronounced "deck"), a double-ended queue
 
 This particular implementation is simple but not very efficient. Several
 operations are O(n). A more efficient implementation would use a doubly
 linked list or a circular buffer.
 */
public struct Deque<T> {
    private var array = [T]()
    
    public var isEmpty: Bool {
        return array.isEmpty
    }
    
    public var count: Int {
        return array.count
    }
    
    public mutating func enqueue(_ element: T) {
        array.append(element)
    }
    
    public mutating func enqueueFront(_ element: T) {
        array.insert(element, at: 0)
    }
    
    public mutating func dequeue() -> T? {
        if isEmpty {
            return nil
        } else {
            return array.removeFirst()
        }
    }
    
    public mutating func dequeueBack() -> T? {
        if isEmpty {
            return nil
        } else {
            return array.removeLast()
        }
    }
    
    public func peekFront() -> T? {
        return array.first
    }
    
    public func peekBack() -> T? {
        return array.last
    }
}
